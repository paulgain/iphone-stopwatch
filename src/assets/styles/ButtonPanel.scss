@import 'core/colours';

@mixin ring($color, $bg-color) {
  color: $color;
  background-color: $bg-color;
  box-shadow: 0 0 0 2px $bg-color;
}

.container {
  display: flex;
  justify-content: space-between;

  button {
    width: 63px;
    height: 63px;
    border-radius: 50%;
    margin: 10px;
    outline: none;
    font-size: 12px;
    letter-spacing: 2px;
    font-weight: 700;
    border: 1px solid $col-app-bg;
    font-family: inherit;
    &.start {
      @include ring($col-btn-fg-start, $col-btn-bg-start);
    }
    &.stop {
      @include ring($col-btn-fg-stop, $col-btn-bg-stop);
    }
    &.reset {
      @include ring($col-btn-fg-lap, $col-btn-bg-lap);
    }
    &:disabled {
      @include ring($col-btn-fg-lap-disabled, $col-btn-bg-lap-disabled);
    }
  }
}
