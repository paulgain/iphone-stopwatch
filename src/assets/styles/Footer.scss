@import 'core/colours';

.footer {
  height: 40px;
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  background-color: $col-app-footer-bg;
}

.icon-container {
  display: flex;
  flex-direction: column;
  align-items: center;

  img {
    padding-top: 2px;
    padding-bottom: 2px;
    width: 24px;
    height: 24px;
    &[alt='Alarm'] {
      padding-top: 5px;
      padding-bottom: 2px;
      width: 21px;
      height: 21px;
    }
  }

  span {
    color: #807f80;
    font-size: 8px;
    font-weight: bolder;
    font-family: Helvetica, serif;
    padding-top: 2px;
    &.selected {
      color: #ff9400;;
    }
  }
}
