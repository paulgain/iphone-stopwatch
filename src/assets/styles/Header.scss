@import 'core/colours';
@import 'core/variables';
@import 'core/mixins';

$image-height: 9px;
$font-size: 9px;
$margin: 3px;
$margin-top: 2px;

.header {
  height: $app-header-height;
  background-color: $col-app-header-bg;
  position: relative;
}

.carrier-and-wifi {
  position: absolute;
  left: 0;
  top: $margin-top;
  @include flex-align-center;
  img {
    height: $image-height;
    margin-left: $margin;
  }
  span {
    font-size: $font-size;
    color: $col-off-white;
    font-family: Helvetica, serif;
    margin-left: $margin;
  }
}

.time {
  @include flex-justify-center;
  span {
    color: $col-off-white;
    font-size: 10px;
    font-weight: 400;
    font-family: Helvetica, serif;
    padding-top: 1px;
  }
}

.bluetooth-and-battery {
  position: absolute;
  right: 0;
  top: $margin-top;
  @include flex-align-center;
  span {
    font-size: $font-size;
    color: $col-off-white;
    font-weight: 700;
    font-family: 'Helvetica Neue', Helvetica, serif;
    margin-right: $margin;
  }
  img {
    height: $image-height;
    margin-right: $margin;
  }
}

.heading {
  @include flex-justify-center;
  span {
    color: white;
    font-size: 14px;
    padding-top: 11px;
    font-family: Helvetica, serif;
  }
}
