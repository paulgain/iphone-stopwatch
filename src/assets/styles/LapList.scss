@import 'core/variables';
@import 'core/colours';

$border-line-colour: #1f1818;

.container {
  height: $app-list-height;
  overflow: hidden;
  overflow-y: scroll;

  &.backgroundLines {
    background-image: linear-gradient($border-line-colour 1px, transparent 1px);
    background-size: 80% 33px;
  }
  &.borderLines {
    dl {
      border-top: 1px solid $border-line-colour;
    }
  }
}
