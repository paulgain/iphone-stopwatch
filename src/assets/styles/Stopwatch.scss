@import 'font';
@import 'core/variables';
@import 'core/colours';

.stopwatch {
    width: $app-width;
    margin: auto;
    display: flex;
    flex-direction: column;
    background-color: $col-app-bg;
}

.name {
  color: white;
}
