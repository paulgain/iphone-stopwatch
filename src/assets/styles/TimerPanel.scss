@import 'core/mixins';
@import 'core/variables';

.container {
  height: $app-time-height;
  @include flex-center;
  span {
    color: white;
    font-size: 70px;
    font-weight: 100;
    padding-top: 11px;
    &.hours {
      font-size: 54px;
    }
  }
}
