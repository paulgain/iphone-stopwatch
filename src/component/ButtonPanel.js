import React from 'react';
import PropTypes from 'prop-types';
import EVENT_TYPE from '../event/EventType';
import Button from './Button';
import '../assets/styles/ButtonPanel.scss';

const createButton = (text, disabled, classNames, onButtonClick) => (
  <Button
    text={text}
    disabled={disabled}
    classNames={classNames}
    onButtonClick={onButtonClick}
  />
);

const ButtonPanel = ({ eventType, onButtonClick }) => {
  let leftButton;
  let rightButton;

  switch (eventType) {
    case EVENT_TYPE.RESET:
      leftButton = createButton('Lap', true, '', onButtonClick);
      rightButton = createButton('Start', false, 'start', onButtonClick);
      break;
    case EVENT_TYPE.START:
      leftButton = createButton('Lap', false, 'reset', onButtonClick);
      rightButton = createButton('Stop', false, 'stop', onButtonClick);
      break;
    case EVENT_TYPE.STOP:
      leftButton = createButton('Reset', false, 'reset', onButtonClick);
      rightButton = createButton('Start', false, 'start', onButtonClick);
      break;
    default: throw new Error(`Unknown event type: ${eventType}`);
  }

  return (
    <div styleName="container">
      { leftButton }
      { rightButton }
    </div>
  );
};

ButtonPanel.propTypes = {
  eventType: PropTypes.string.isRequired,
  onButtonClick: PropTypes.func.isRequired
};

export default ButtonPanel;
