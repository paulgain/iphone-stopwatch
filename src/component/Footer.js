import React from 'react';
import PropTypes from 'prop-types';
import bed from '../assets/img/bed.svg';
import globe from '../assets/img/globe.svg';
import alarm from '../assets/img/alarm.svg';
import timer from '../assets/img/timer.svg';
import stopWatch from '../assets/img/stop-watch.svg';
import '../assets/styles/Footer.scss';

const Icon = ({ svg, text, selected }) => (
  <div styleName="icon-container">
    <img src={svg} alt={text} />
    <span styleName={selected ? 'selected' : ''}>{text}</span>
  </div>
);

const Footer = () => (
  <div styleName="footer">
    <Icon svg={globe} text="World Clock" />
    <Icon svg={alarm} text="Alarm" />
    <Icon svg={bed} text="Bedtime" />
    <Icon svg={stopWatch} text="Stopwatch" selected />
    <Icon svg={timer} text="Timer" />
  </div>
);

Icon.defaultProps = {
  selected: false
};

Icon.propTypes = {
  svg: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  selected: PropTypes.bool
};

export default Footer;
