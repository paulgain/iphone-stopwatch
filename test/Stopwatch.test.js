import React from 'react';
import TestRenderer from 'react-test-renderer';
import Stopwatch from '../src/component/Stopwatch';
import Header from '../src/component/Header';
import ButtonPanel from '../src/component/ButtonPanel';
import TimerPanel from '../src/component/TimerPanel';
import LapList from '../src/component/LapList';
import Footer from '../src/component/Footer';
import Timer from '../src/time/Timer';
import EVENT_TYPE from '../src/event/EventType';

describe('Stopwatch', () => {
  const initialState = {
    eventType: 'Reset',
    time: 0,
    listTime: 0,
    lapTimes: []
  };

  describe('static initialState()', () => {
    test('it should create the state', () => {
      expect(Stopwatch.initialState()).toEqual(initialState);
    });
  });

  describe('constructor()', () => {
    test('should define 2 timers, state, props and requestID', () => {
      const testRenderer = TestRenderer.create(<Stopwatch />);
      const instance = testRenderer.getInstance();
      expect(instance.timer).toBeInstanceOf(Timer);
      expect(instance.listTimer).toBeInstanceOf(Timer);
      expect(instance.props).toEqual({});
      expect(instance.state).toEqual(initialState);
      expect(instance.handle).not.toBe(null);
    });
  });

  describe('onButtonClick(event)', () => {
    const stopwatch = new Stopwatch();
    const event = { target: { name: 'Start' } };

    test('it should call the start() function', () => {
      stopwatch.start = jest.fn();
      stopwatch.onButtonClick(event);
      expect(stopwatch.start).toHaveBeenCalled();
    });

    test('it should call the stop() function', () => {
      stopwatch.stop = jest.fn();
      event.target.name = 'Stop';
      stopwatch.onButtonClick(event);
      expect(stopwatch.stop).toHaveBeenCalled();
    });

    test('it should call the lap() function', () => {
      stopwatch.lap = jest.fn();
      event.target.name = 'Lap';
      stopwatch.onButtonClick(event);
      expect(stopwatch.lap).toHaveBeenCalled();
    });

    test('it should call the reset() function', () => {
      stopwatch.reset = jest.fn();
      event.target.name = 'Reset';
      stopwatch.onButtonClick(event);
      expect(stopwatch.reset).toHaveBeenCalled();
    });

    test('it should throw an error', () => {
      expect(() => stopwatch.onButtonClick({ target: { name: 'Doh' } }))
        .toThrowError('Unknown click event: Doh');
    });
  });

  describe('stopwatch actions start(), stop(), reset() and lap()', () => {
    let stopwatch;

    beforeEach(() => {
      stopwatch = new Stopwatch();
    });

    it('it should start both timers and update the state', () => {
      stopwatch.timer = { start: jest.fn() };
      stopwatch.listTimer = { start: jest.fn() };
      stopwatch.setState = jest.fn();
      stopwatch.start();

      expect(stopwatch.timer.start).toHaveBeenCalled();
      expect(stopwatch.listTimer.start).toHaveBeenCalled();
      expect(stopwatch.setState).toHaveBeenCalledWith({ eventType: EVENT_TYPE.START });
    });

    it('it should stop both timers and update the state', () => {
      stopwatch.timer = { stop: jest.fn() };
      stopwatch.listTimer = { stop: jest.fn() };
      stopwatch.setState = jest.fn();
      stopwatch.stop();

      expect(stopwatch.timer.stop).toHaveBeenCalled();
      expect(stopwatch.listTimer.stop).toHaveBeenCalled();
      expect(stopwatch.setState).toHaveBeenCalledWith({ eventType: EVENT_TYPE.STOP });
    });

    it('it should reset both timers and update the state', () => {
      stopwatch.timer = { reset: jest.fn() };
      stopwatch.listTimer = { reset: jest.fn() };
      stopwatch.setState = jest.fn();
      stopwatch.reset();

      expect(stopwatch.timer.reset).toHaveBeenCalled();
      expect(stopwatch.listTimer.reset).toHaveBeenCalled();
      expect(stopwatch.setState).toHaveBeenCalledWith(Stopwatch.initialState());
    });

    it('it should save the current lap time', () => {
      stopwatch.createLapTime = jest.fn();
      stopwatch.listTimer = { reset: jest.fn(), start: jest.fn() };
      stopwatch.setState = jest.fn();
      stopwatch.lap();

      expect(stopwatch.createLapTime).toHaveBeenCalled();
      expect(stopwatch.listTimer.reset).toHaveBeenCalled();
      expect(stopwatch.listTimer.start).toHaveBeenCalled();
      expect(stopwatch.setState).toHaveBeenCalledWith(stopwatch.addLapTime);
    });
  });

  describe('helper functions', () => {
    let stopwatch;

    beforeEach(() => {
      stopwatch = new Stopwatch();
      stopwatch.listTimer = { time: jest.fn().mockReturnValue(100) };
      stopwatch.createLapTime();
    });

    it('it should create a lap time', () => {
      expect(stopwatch.currentLapTime.id).toBeDefined();
      expect(stopwatch.currentLapTime.lapTime).toEqual(100);
    });

    it('it should immutably add a lap time to the previous lap times', () => {
      const prevState = { lapTimes: [{ id: 'UTWe8YB9Cs', lapTime: 10 }] };
      const newState = stopwatch.addLapTime(prevState);

      expect(newState.lapTimes.length).toEqual(2);

      // Test for immutability
      expect(newState).not.toBe(prevState);
      expect(newState.lapTimes).not.toBe(prevState.lapTimes);
    });
  });

  describe('stopwatch UI sections', () => {
    const { root } = TestRenderer.create(<Stopwatch />);

    test('it should contain a Header', () => {
      const header = root.findByType(Header);
      expect(header.props).toEqual({});
    });

    test('it should contain a TimerPanel', () => {
      const timerPanel = root.findByType(TimerPanel);
      expect(timerPanel.props.time).toBe(0);
    });

    test('it should contain a ButtonPanel', () => {
      const buttonPanel = root.findByType(ButtonPanel);
      expect(buttonPanel.props.eventType).toBe('Reset');
      expect(buttonPanel.props.onButtonClick).toBeInstanceOf(Function);
    });

    test('it should contain a LapList', () => {
      const lapList = root.findByType(LapList);
      expect(lapList.props.listTime).toBe(0);
      expect(lapList.props.lapTimes).toEqual([]);
      expect(lapList.props.eventType).toBe('Reset');
    });

    test('it should contain a Footer', () => {
      const tooter = root.findByType(Footer);
      expect(tooter.props).toEqual({});
    });
  });

  describe('Snapshot', () => {
    test('it should match the local snapshot', () => {
      const testRenderer = TestRenderer.create(<Stopwatch />);
      expect(testRenderer.toJSON()).toMatchSnapshot();
    });
  });
});
