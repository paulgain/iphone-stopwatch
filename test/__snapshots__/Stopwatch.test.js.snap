// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stopwatch it should match the local snapshot 1`] = `
<div
  styleName="stopwatch"
>
  <div
    styleName="header"
  >
    <div
      styleName="carrier-and-wifi"
    >
      <img
        alt="phone signal"
        src="test-file-stub"
      />
      <span>
        3
      </span>
      <img
        alt="wifi connection"
        src="test-file-stub"
      />
    </div>
    <div
      styleName="time"
    >
      <span>
        00.15
      </span>
    </div>
    <div
      styleName="bluetooth-and-battery"
    >
      <img
        alt="bluetooth connection"
        src="test-file-stub"
      />
      <span>
        25%
      </span>
      <img
        alt="battery life"
        src="test-file-stub"
      />
    </div>
    <div
      styleName="heading"
    >
      <span>
        Stopwatch
      </span>
    </div>
  </div>
  <div
    styleName="container"
  >
    <span
      styleName=""
    >
      00:00:00
    </span>
  </div>
  <div
    styleName="container"
  >
    <button
      disabled={true}
      name="Lap"
      onClick={[Function]}
      styleName=""
      type="button"
    >
      Lap
    </button>
    <button
      disabled={false}
      name="Start"
      onClick={[Function]}
      styleName="start"
      type="button"
    >
      Start
    </button>
  </div>
  <div
    styleName="container backgroundLines"
  />
  <div
    styleName="footer"
  >
    <div
      styleName="icon-container"
    >
      <img
        alt="World Clock"
        src="test-file-stub"
      />
      <span
        styleName=""
      >
        World Clock
      </span>
    </div>
    <div
      styleName="icon-container"
    >
      <img
        alt="Alarm"
        src="test-file-stub"
      />
      <span
        styleName=""
      >
        Alarm
      </span>
    </div>
    <div
      styleName="icon-container"
    >
      <img
        alt="Bedtime"
        src="test-file-stub"
      />
      <span
        styleName=""
      >
        Bedtime
      </span>
    </div>
    <div
      styleName="icon-container"
    >
      <img
        alt="Stopwatch"
        src="test-file-stub"
      />
      <span
        styleName="selected"
      >
        Stopwatch
      </span>
    </div>
    <div
      styleName="icon-container"
    >
      <img
        alt="Timer"
        src="test-file-stub"
      />
      <span
        styleName=""
      >
        Timer
      </span>
    </div>
  </div>
</div>
`;
